### ISSUE TITLE
Optional username/channel treated differently on different OSes

### ISSUE THREAD

**Issue by Artalus** [conan-io/conan-package-tools#479]
We attempt to use Github Actions + CPT in Conan-Center-Index fork. Most of recipes there use new convention with optional username and channel for packages, and it seems CPT is not quite ready for this:

- if we specify `CONAN_REFERENCE` as simply `lib/ver`, CPT on Linux fails with this:
```
Traceback (most recent call last):
  File "/home/iivanov/git/conan-center-index/build.py", line 6, in <module>
    builder = ConanMultiPackager()
  File "/home/iivanov/.local/lib/python3.8/site-packages/cpt/packager.py", line 211, in __init__
    self.reference = ConanFileReference(name, version, self.username, self.channel)
  File "/home/iivanov/.local/lib/python3.8/site-packages/conans/model/ref.py", line 164, in __new__
    raise InvalidNameException("Specify the 'user' and the 'channel' or neither of them")
conans.errors.InvalidNameException: Specify the 'user' and the 'channel' or neither of them
```
This seems to happen due to `channel` being set to `testing` automatically in CPT innards.

- if we [change it to `lib/ver@/`](https://github.com/trassir/conan-center-index/pull/7/commits/38ada2c1d4ba2d26a7d5f3bd6b197c0da6d0b19f), CPT [works fine on Linux](https://github.com/trassir/conan-center-index/runs/486995222#step:5:372), but will [fail on Mac](https://github.com/trassir/conan-center-index/runs/486995245#step:5:211)
```
Traceback (most recent call last):
  File "../../../build.py", line 8, in <module>
    builder.run()
  File "/Users/runner/hostedtoolcache/Python/3.8.1/x64/lib/python3.8/site-packages/cpt/packager.py", line 541, in run
    self.run_builds(base_profile_name=base_profile_name)
  File "/Users/runner/hostedtoolcache/Python/3.8.1/x64/lib/python3.8/site-packages/cpt/packager.py", line 633, in run_builds
    r.run()
  File "/Users/runner/hostedtoolcache/Python/3.8.1/x64/lib/python3.8/site-packages/cpt/runner.py", line 121, in run
    self._results = self._conan_api.create(self._conanfile, name=name, version=version,
  File "/Users/runner/hostedtoolcache/Python/3.8.1/x64/lib/python3.8/site-packages/conans/client/conan_api.py", line 81, in wrapper
    return f(api, *args, **kwargs)
  File "/Users/runner/hostedtoolcache/Python/3.8.1/x64/lib/python3.8/site-packages/conans/client/conan_api.py", line 341, in create
    new_ref = cmd_export(self.app, conanfile_path, name, version, user, channel, keep_source,
  File "/Users/runner/hostedtoolcache/Python/3.8.1/x64/lib/python3.8/site-packages/conans/client/cmd/export.py", line 80, in cmd_export
    ref = ConanFileReference(conanfile.name, conanfile.version, user, channel)
  File "/Users/runner/hostedtoolcache/Python/3.8.1/x64/lib/python3.8/site-packages/conans/model/ref.py", line 164, in __new__
    raise InvalidNameException("Specify the 'user' and the 'channel' or neither of them")
conans.errors.InvalidNameException: Specify the 'user' and the 'channel' or neither of them
```

- if, however, we specify envvars `CONAN_USERNAME=_` and `CONAN_CHANNEL=_`, then [Linux fails](https://github.com/trassir/conan-center-index/runs/487019691#step:5:347), yet [Mac works](https://github.com/trassir/conan-center-index/runs/487019722#step:5:216)
```
Traceback (most recent call last):
  File "/opt/pyenv/versions/3.6.7/bin/run_create_in_docker", line 11, in <module>
    load_entry_point('conan-package-tools==0.31.1', 'console_scripts', 'run_create_in_docker')()
  File "/opt/pyenv/versions/3.6.7/lib/python3.6/site-packages/cpt/run_in_docker.py", line 53, in run
    runner.run()
  File "/opt/pyenv/versions/3.6.7/lib/python3.6/site-packages/cpt/runner.py", line 127, in run
    update=self._update_dependencies)
  File "/opt/pyenv/versions/3.6.7/lib/python3.6/site-packages/conans/client/conan_api.py", line 81, in wrapper
    return f(api, *args, **kwargs)
  File "/opt/pyenv/versions/3.6.7/lib/python3.6/site-packages/conans/client/conan_api.py", line 343, in create
    ignore_dirty=ignore_dirty)
  File "/opt/pyenv/versions/3.6.7/lib/python3.6/site-packages/conans/client/cmd/export.py", line 80, in cmd_export
    ref = ConanFileReference(conanfile.name, conanfile.version, user, channel)
  File "/opt/pyenv/versions/3.6.7/lib/python3.6/site-packages/conans/model/ref.py", line 164, in __new__
    raise InvalidNameException("Specify the 'user' and the 'channel' or neither of them")
conans.errors.InvalidNameException: Specify the 'user' and the 'channel' or neither of them
Traceback (most recent call last):
  File "/home/iivanov/git/conan-center-index/build.py", line 8, in <module>
    builder.run()
  File "/home/iivanov/.local/lib/python3.8/site-packages/cpt/packager.py", line 547, in run
    self.run_builds(base_profile_name=base_profile_name)
  File "/home/iivanov/.local/lib/python3.8/site-packages/cpt/packager.py", line 673, in run_builds
    r.run(pull_image=not pulled_docker_images[docker_image],
  File "/home/iivanov/.local/lib/python3.8/site-packages/cpt/runner.py", line 304, in run
    raise Exception("Error building: %s" % command)
```

The [current working solution](https://github.com/trassir/conan-center-index/pull/7/commits/6d5da0e873e2b6878a112292bcb8753c385de97c) for us is to specify `CONAN_REFERENCE` for Linux builds and `CONAN_CHANNEL+_USERNAME` for Mac builds, but this is rather cumbersome.

### Environment Details
* Conan Package Tools Version: **0.31.1**
* Operating System: **Linux / Mac**
* Conan version: **conan 1.22.2**

**Comment by Artalus** [comment]
On a side note, it also seems that when no `CONAN_STABLE_CHANNEL` is provided, CPT will [substitute it for stable](https://github.com/conan-io/conan-package-tools/blob/develop/cpt/packager.py#L203). I am unsure if this needs any correction, since omitting username and channel have not yet become a widespread practice.

**Comment by Artalus** [comment]
Also, with `CONAN_STABLE_CHANNEL=_` one will also get an error saying
```
 >> Branch detected
   >> trassir-ci
 >> Info
   >> Redefined channel by CI branch matching with 'trassir-ci', setting CONAN_CHANNEL to '_'
Traceback (most recent call last):
...
conans.errors.InvalidNameException: Specify the 'user' and the 'channel' or neither of them
```

**Comment by Artalus** [comment]
So I did some more digging and turns out the problem is not Linux-specific, but Docker-specific. The issue was `CONAN_USERNAME` being discarded instead of forwarding to Docker. I made a PR fixing this.

**Comment by uilianries** [comment]
Fixed by #522

Should be available on CPT 0.35.1
